lib/ecto/repo/preloader.ex

```diff
@@ -113,52 +113,65 @@ defmodule Ecto.Repo.Preloader do
 
   defp preload_assoc(structs, module, repo, prefix, %{cardinality: card} = assoc,
                      related_key, query, preloads, take, opts) do
-    {fetch_ids, loaded_ids, loaded_structs} =
-      fetch_ids(structs, module, assoc, opts)
+    {fetch_structs, loaded_ids, loaded_structs} =
+      fetch_structs(structs, module, assoc, opts)
+
+    fetch_structs = if is_function(query, 1) do
+                      fetch_query(structs, assoc, query, prefix, related_key, take)
+                    else
+                      assoc.__struct__.preload(assoc, repo, query, take, prefix, fetch_structs, opts)
+                    end
+
     {fetch_ids, fetch_structs} =
-      fetch_query(fetch_ids, assoc, repo, query, prefix, related_key, take, opts)
+      fetch_structs
+      |> unzip_structs([], [])
 
     all = preload_each(Enum.reverse(loaded_structs, fetch_structs), repo, preloads, opts)
     {:assoc, assoc, assoc_map(card, Enum.reverse(loaded_ids, fetch_ids), all)}
   end
 
-  defp fetch_ids(structs, module, assoc, opts) do
+  defp fetch_structs(structs, module, assoc, opts) do
     %{field: field, owner_key: owner_key, cardinality: card} = assoc
     force? = Keyword.get(opts, :force, false)
 
     Enum.reduce structs, {[], [], []}, fn
       nil, acc ->
         acc
-      struct, {fetch_ids, loaded_ids, loaded_structs} ->
+      struct, {fetch_structs, loaded_ids, loaded_structs} ->
```

Comment:
Can you please explain why you have to return structs now?

Reply:
So from what I could understand this function divides up the ones that are already loaded from the ones we need to fetch. If I didn't return the structs that need fetching and left this function alone then I'd be passing in `structs` (which is all loaded and needing to be fetched) which would end up fetching an association that was already loaded.